uvloop/handles/udp.pyx

```diff
@@ -1,199 +1,373 @@
+        uv.uv_buf_t     uv_buf
+        Py_buffer       py_buf
 
+        UDPTransport    udp
 
-cdef class UDPTransport(UVBaseTransport):
+        bint            closed
+
+    cdef close(self):
+        if self.closed:
+            return
+
+        self.closed = 1
+        PyBuffer_Release(&self.py_buf)  # void
+        self.req.data = NULL
+        Py_DECREF(self)
+        self.udp = None
 
+    @staticmethod
+    cdef _UDPSendContext new(UDPTransport udp, object data):
+        cdef _UDPSendContext ctx
+        ctx = _UDPSendContext.__new__(_UDPSendContext)
+        ctx.udp = None
+        ctx.closed = 1
+
+        ctx.req.data = <void*> ctx
+        Py_INCREF(ctx)
+
+        PyObject_GetBuffer(data, &ctx.py_buf, PyBUF_SIMPLE)
+        ctx.uv_buf = uv.uv_buf_init(<char*>ctx.py_buf.buf, ctx.py_buf.len)
+        ctx.udp = udp
+
+        ctx.closed = 0
+        return ctx
+
+    def __dealloc__(self):
+        if UVLOOP_DEBUG:
+            if not self.closed:
+                raise RuntimeError(
+                    'open _UDPSendContext is being deallocated')
+        self.udp = None
+
+
+@cython.no_gc_clear
+cdef class UDPTransport(UVBaseTransport):
     def __cinit__(self):
-        self.sock = None
-        self.poll = None
-        self.buffer = col_deque()
-        self._has_handle = 0
+        self._family = uv.AF_UNSPEC
+        self.__receiving = 0
+
+    cdef _init(self, Loop loop, unsigned int family):
+        cdef int err
 
-    cdef _init(self, Loop loop, object sock, object r_addr):
         self._start_init(loop)
-        try:
-            # It's important to incref the socket in case it
-            # was created outside of uvloop,
-            # i.e. `look.create_datagram_endpoint(sock=sock)`.
-            socket_inc_io_ref(sock)
-
-            self.sock = sock
-            self.sock_family = sock.family
-            self.sock_proto = sock.proto
-            self.sock_type = sock.type
-            self.address = r_addr
-            self.poll = UVPoll.new(loop, sock.fileno())
-            self._finish_init()
-        except Exception:
-            self._free()
+
+        self._handle = <uv.uv_handle_t*>PyMem_RawMalloc(sizeof(uv.uv_udp_t))
+        if self._handle is NULL:
             self._abort_init()
-            raise
+            raise MemoryError()
 
-    cdef size_t _get_write_buffer_size(self):
-        cdef int size = 0
-        for data, addr in self.buffer:
-            size += len(data)
-        return size
+        err = uv.uv_udp_init_ex(loop.uvloop,
+                                <uv.uv_udp_t*>self._handle,
+                                family)
+        if err < 0:
+            self._abort_init()
+            raise convert_error(err)
+
+        if family in (uv.AF_INET, uv.AF_INET6):
+            self._family = family
+
+        self._finish_init()
+
+    cdef _connect(self, system.sockaddr* addr, size_t addr_len):
+        cdef int err
+        uv.uv_udp_connect(<uv.uv_udp_t*>self._handle, addr)
```

Comment:
It looks like `err` should be assigned?

Reply:
Yeah, good catch! I've pushed a commit with a fix.